pkg/destroy/libvirt/connection/connection.go

```diff
@@ -0,0 +1,111 @@
+//go:build libvirt
+// +build libvirt
+
+package connection
+
+import (
+	"crypto/tls"
+	"fmt"
+	"net"
+	"net/url"
+	"strings"
+	"time"
+
+	libvirt "github.com/digitalocean/go-libvirt"
+	"github.com/sirupsen/logrus"
+)
+
+// NewConnection creates a new connection to a libvirt daemon.
+func NewConnection(libvirtURI string, log logrus.FieldLogger) (conn *libvirt.Libvirt, disconnect func(), err error) {
+	disconnect = func() {
+		if conn != nil {
+			if err := conn.Disconnect(); err != nil && log != nil {
+				log.Warn("failed to disconnect from libvirt", err)
+			}
+		}
+	}
+
+	uri, err := url.Parse(libvirtURI)
+	if err != nil {
+		return
+	}
+
+	c, err := dial(uri)
+	if err != nil {
+		err = fmt.Errorf("could not open connection to libvirt: %w", err)
+		return
+	}
+
+	conn = libvirt.New(c)
+	err = connect(uri, conn, log)
+	return
+}
+
+func dial(uri *url.URL) (net.Conn, error) {
+	transport := "unix"
+	if scheme := strings.Split(uri.Scheme, "+"); len(scheme) > 1 {
+		transport = scheme[1]
+	} else if uri.Host != "" {
+		transport = "tls"
+	}
+
+	switch transport {
+	case "unix":
+		sock := "/var/run/libvirt/libvirt-sock"
+		if s := uri.Query().Get("socket"); s != "" {
+			sock = s
+		}
+		return net.DialTimeout(transport, sock, 2*time.Second)
+	case "tcp":
+		host := uri.Host
+		if uri.Port() == "" {
+			host = net.JoinHostPort(uri.Hostname(), "16509")
+		}
+		return net.DialTimeout(transport, host, 2*time.Second)
+	case "tls":
+		host := uri.Host
+		if uri.Port() == "" {
+			host = net.JoinHostPort(uri.Hostname(), "16514")
+		}
+		config, err := tlsConfig(uri)
+		if err != nil {
+			return nil, err
+		}
+		c, err := tls.Dial("tcp", host, config)
+		if err == nil {
+			// https://github.com/digitalocean/go-libvirt/issues/89
```

Comment:
https://github.com/digitalocean/go-libvirt/issues/89#issuecomment-1607300636 in particular has a good explanation for it.